Fix state-machine in hydra scripts

ch1bo opened the pull request:
When reviewing the [hydra-spec](https://www.overleaf.com/read/bbqzmptcxryj) lately, I discovered three short-comings still in our scripts:
* We did not check contract continuity, i.e. that the output with the state datum and ST is actually owned by vHead
* vHead / collectCom did not actually check whether all the value is collected
* vInitial did not check that the right headId is in the output

:cat2: Fixes the short-comings above and checking them using new or updated mutation tests.

:leopard: ~~This PR also adds and uses [plutonomy](https://github.com/well-typed/plutonomy/tree/master/src/Plutonomy) to decrease script sizes (to fit publish-scripts tx again) and improve performance of the generated UPLC.~~ Use custom script contexts in initial and commit script validators to reduce tx size.

For reviewing this, the relevant changes are marked red in the [spec](https://www.overleaf.com/read/bbqzmptcxryj).

---

<!-- Tick off or strike-through / remove if not applicable -->
* [x] CHANGELOG updated
* [x] Documentation update not needed
* [x] Added and/or updated haddocks
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Identify short-comings of collect, close & contest

The scripts do not ensure that contract state is paid to the head
validator again. Also, collect has no explicit check of the collected
value.

ch1bo committed:
Re-indent Contest mutations

There was an indentation error and I commit it separately to make the
change more clearly visible.

ch1bo committed:
Add mutations to fail collect, close and contest

The vHead validator does not ensure that the state is retained in vHead.

ch1bo committed:
Ensure head scripts pay to head

By checking for the address when we ind the head output datum, we ensure
we only consider the first output which must be addressed to vHead.

ch1bo committed:
Fix vInitial to check head id in output datum

This needs to be ensured, because if it does not match, the Head is
locked and cannot be spent in an abort or collectCom transaction.

ch1bo committed:
Fix CollectCom to ensure it collects all value

We seemingly have dropped this technique accidentially. Instead of
checking the head output value against all the collected value, we would
check the other inputs & outputs balance. This is more efficient as we
usually have a more constant number of inputs and outputs to pay for
fees / get change in return.

ch1bo committed:
Randomize removed value from collectCom mutation

ch1bo committed:
Remove redundant FIXME and add an XXX in HeadLogic

ch1bo committed:
Update changelog and version of hydra-plutus

v0d1ch commented:
@ch1bo Seems like cardano-node thread just dies causing a timeout:
```
{"app":[],"at":"2023-03-28T08:58:55.19Z","data":{"kind":"TraceImmutableDBEvent.DBClosed"},"env":"1.35.4:00000","host":"nixos","loc":null,"msg":"","ns":["cardano.node.ChainDB"],"pid":"103252","sev":"Info","thread":"5"}
{"app":[],"at":"2023-03-28T08:58:55.19Z","data":{"immtip":{"kind":"GenesisPoint"},"kind":"TraceOpenEvent.ClosedDB","tip":{"headerHash":"20904a9ab326e76302302adf1dd01e6857101c9d2497d5a66fe678bcf5aee0c1","kind":"BlockPoint","slot":591}},"env":"1.35.4:00000","host":"nixos","loc":null,"msg":"","ns":["cardano.node.ChainDB"],"pid":"103252","sev":"Info","thread":"5"}
cardano-node: thread killed

```

v0d1ch reviewed (approved):
LGTM!

ch1bo commented:
We should have another look why the timeout increase was necessary. The tests where we increased it must have something in common.

ch1bo commented:
I realized that the `hydra-node` startup became slower through this change. As @pgrange pointed out, it could be because of `plutonomy`. And indeed, `Plutonomy.optimizeUPLC` is an expensive function to run and it would be invoked on first use of our validators .. afterwards the result is shared by Haskell's value sharing system.

We also discussed today that we want to investigate other avenues alternative to plutonomy to get this merged.

ch1bo committed:
Create a custom Hydra.ScriptContext and use it in Commit & Initial validator

ch1bo committed:
Regenerate golden plutus scripts

Obviously the script fixes resul in new scripts / hashes.

ch1bo committed:
Remove two FIXMEs

One was already addressed (see mutation tests), the other was more an
information for future uses of the custom script context.

v0d1ch commented on hydra-plutus/src/Hydra/Contract/Head.hs:
```diff
@@ -146,33 +146,26 @@ checkAbort ctx@ScriptContext{scriptContextTxInfo = txInfo} headCurrencySymbol pa
+  mustCollectUtxoHash
     && mustNotChangeParameters
+    && mustCollectAllValue
+    -- XXX: Is this really needed? If yes, why not check on the output?
```
I've tried checking both inputs and outputs and it doesen't seem to increase the tx costs much. Here is the diff:
```diff
diff --git a/hydra-plutus/src/Hydra/Contract/Head.hs b/hydra-plutus/src/Hydra/Contract/Head.hs
index 519206b6d..1d344fc5b 100644
--- a/hydra-plutus/src/Hydra/Contract/Head.hs
+++ b/hydra-plutus/src/Hydra/Contract/Head.hs
@@ -44,7 +44,7 @@ import Plutus.V2.Ledger.Api (
   adaToken,
   mkValidatorScript,
  )
-import Plutus.V2.Ledger.Contexts (findDatum, findOwnInput)
+import Plutus.V2.Ledger.Contexts (findDatum, findOwnInput, getContinuingOutputs)
 import PlutusTx (CompiledCode)
 import qualified PlutusTx
 import qualified PlutusTx.AssocMap as Map
@@ -161,8 +161,8 @@ checkCollectCom ctx@ScriptContext{scriptContextTxInfo = txInfo} (contestationPer
   mustCollectUtxoHash
     && mustNotChangeParameters
     && mustCollectAllValue
-    -- XXX: Is this really needed? If yes, why not check on the output?
-    && traceIfFalse $(errorCode STNotSpent) (hasST headId val)
+    && traceIfFalse $(errorCode STNotSpent) (hasST headId inval)
+    && traceIfFalse $(errorCode STNotSpent) (hasST headId outval)
     && everyoneHasCommitted
     && mustBeSignedByParticipant ctx headId
     && mustNotMintOrBurn txInfo
@@ -204,7 +204,8 @@ checkCollectCom ctx@ScriptContext{scriptContextTxInfo = txInfo} (contestationPer
     traceIfFalse $(errorCode MissingCommits) $
       nTotalCommits == length parties
 
-  val = maybe mempty (txOutValue . txInInfoResolved) $ findOwnInput ctx
+  inval = maybe mempty (txOutValue . txInInfoResolved) $ findOwnInput ctx
+  outval = foldMap txOutValue (getContinuingOutputs ctx)
 
   otherValueOut =
     case txInfoOutputs txInfo of

```

ch1bo replied:
Yeah I'm not so sure if I want to do it double. I think we should decide what is more intuitive. I'll ask tomorrow in the tactical.

ch1bo replied:
Decision: Checking for the ST should be part of the `mustCollectAllValue` rule (to match the spec) and we hence should check that there is an ST in the output.

Conretely, this means that we need to put a `hasST` check to `mustCollectAllValue` because the current clause (the inverse of the thing in the spec) does not cover it yet.

v0d1ch commented on hydra-plutus/src/Hydra/ScriptContext.hs:
```diff
@@ -0,0 +1,118 @@
+-- | The 'CurrencySymbol' of the current validator script.
+ownCurrencySymbol :: ScriptContext -> CurrencySymbol
+ownCurrencySymbol ScriptContext{scriptContextPurpose = Minting cs} = cs
+ownCurrencySymbol _ = traceError "Lh" -- "Can't get currency symbol of the current validator script"
```
Leftover comment?

ch1bo replied:
This is from the original function. I left it unchanged. We might introduce our error codes here at some point.

v0d1ch reviewed (approved):
Nice! I added a diff related to checking also the head outputs for `ST`, perhaps we need to do it in all head transitions but not really mandatory right now.

ffakenz commented on hydra-node/test/Hydra/Chain/Direct/Contract/Mutation.hs:
```diff
@@ -265,8 +265,10 @@ data Mutation
+  | -- | Adds given output as first transaction output.
     PrependOutput (TxOut CtxTx)
+  | -- | Adds given output as last transaction output.
+    AddOutput (TxOut CtxTx)
```
I would name it AppendOutput :P

ffakenz commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -674,7 +670,7 @@ onOpenNetworkReqSn env ledger st otherParty sn requestedTxs =
     -- Spec: wait s̅ = ŝ
     waitNoSnapshotInFlight $
-      -- Spec: wait U̅ ◦ T ̸= ⊥ combined with Û ← Ū̅ ◦ T
+      -- Spec: wait U̅ ◦ Tres = ⊥ combined with Û ← Ū̅ ◦ T
```
what is Tres and why now the condition changed from `/=` to `=`?

Also at the end of the line there is another mention of T, is it not the same?

ch1bo replied:
That was a typo/leftover .. will fix it.
```suggestion
      -- Spec: wait U̅ ◦ T /= ⊥ combined with Û ← Ū̅ ◦ T
```

ch1bo committed:
Rename AddOutput -> AppendOutput and fix a typo
